lib/filestorage.js

```diff
@@ -0,0 +1,156 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+const crypto = require('crypto');
+const common = require('metarhia-common');
+const { crc32 } = require('crc');
+const { zip, gzip } = require('compressing');
+
+const FS_EXT = 'f';
+
+class FileStorage {
+
+  // Create new FileStorage
+  //   dir - <string>, data storage directory
+  constructor(dir) {
+    this.dir = path.resolve(dir);
+    this.compressionList = [];
+    if (!fs.existsSync(dir)) fs.mkdirSync(dir);
+  }
+
+  _computeHash(data, checksum) {
+    switch (checksum) {
+      case 'CRC32':
+        return crc32(data).toString(16);
+      case 'SHA256':
+        return crypto.createHash('sha256')
+          .update(data)
+          .digest('hex');
+      default:
+        throw new TypeError(`${checksum} is not supported`);
+    }
+  }
+
+  _mkdir(dir, cb) {
+    fs.access(dir, fs.constants.F_OK, err => {
+      if (err) {
+        this._mkdir(path.dirname(dir), err => {
+          if (err) cb(err);
+          else fs.mkdir(dir, cb);
+        });
+      } else {
+        cb(null);
+      }
+    });
+  }
+
+  _prepareFilepath(id, cb) {
+    const chunks = common.idToChunks(id);
+    const file = path.join(this.dir, ...chunks) + '.' + FS_EXT;
+    this._mkdir(path.dirname(file), err => {
+      cb(err, file);
+    });
+  }
+
+  // Write file to storage
+  //   id - <Object>, common.Uint64, id of file
+  //   data - <string> | <Buffer> | <Uint8Array>, data to be written
+  //   opts - <Object>
+  //     checksum - <string>, checksum type
+  //     dedupHash - <string>, second checksum type
+  //   cb - <Function>, callback
+  //     err - <error>
+  //     checksum - <string>, data checksum
+  //     dedupHash - <string>, second data checksum
+  //     originalSize - <number>, data size
+  // Throws: <TypeError>, if `opts.checksum` or `opts.dedupHash` is incorrect
+  write(id, data, opts = {}, cb) {
+    const checksum = this._computeHash(data, opts.checksum || 'CRC32');
+    const dedupHash = this._computeHash(data, opts.dedupHash || 'SHA256');
+    const originalSize = data.length;
+
+    this._prepareFilepath(id, (err, file) => {
+      if (err) {
+        cb(err);
+      } else {
+        fs.writeFile(file, data, err => {
+          cb(err, checksum, dedupHash, originalSize);
+        });
+      }
+    });
+  }
+
+  // Read file from storage
+  //   id - <Object>, common.Uint64, id of file
+  //   cb - <Function>, callback
+  //     err - <error>
+  //     data - <Buffer>
+  read(id, cb) {
+    const file = path.join(this.dir, common.idToPath(id) + '.' + FS_EXT);
```

Comment:
Can we extract name creation into a function?
```js
_makeNameExt(name) {
  return `${name}.${FS_EXT}`;
}
```
Name of the function is up to you.